_posts/grant_guo/2018-04-19-kf-etl-pipeline-refactoring.md

```diff
@@ -0,0 +1,212 @@
+---
+layout: post
+title:  "Prorammable Pipeline in Kids-First ETL"
+breadcrumb: true
+author: grant_guo
+date: 2018-04-19
+categories: grant_guo
+tags:
+    - Functional Programming(FP), Scala
+teaser:
+    info: Refactored Kids-First ETL to make the Pipeline programmable
+    image: watch_gears.jpg
+header:
+    version: small
+    title: Software Engineering Blog
+    image: header-logo-crop.png
+    icon: icon-blog
+---
+
+## Forword
+
+In my previous post [Introducing the Kids-First ETL](http://softeng.oicr.on.ca/grant_guo/2018/03/19/kf-etl/), I mentioned:
+
+> A nice future enhacement would be to **make the Pipeline programmable by defining operators** within Pipeline that accept Processor(s) as input and determine how to execute the Processor(s)
+
+In this blog, Let's explore how to do it.
+
+## The flaws of the current implementation
+
+Let's take a look at how the **Pipeline** is implemented:
+
+```
+object Pipeline {
+    private lazy val injector = createInjector()
+    private def createInjector(): Injector = { ... }
+    def run():Unit = {
+        val download = injector.getInstance(classOf[DownloadProcessor])
+        val filecentric = injector.getInstance(classOf[FileCentricProcessor])
+        val participantcentric = injector.getInstance(classOf[ParticipantCentricProcessor])
+        val index = injector.getInstance(classOf[IndexProcessor])
+    
+        val dump_location = download.process()
+        
+        val fp = filecentric.process _
+        fp.andThen(index.process)(dump_location)
+    
+        val pp = participantcentric.process _
+        pp.andThen(index.process)(dump_location)
+    
+    }
+}
+
+```
+
+* Initialize all of the ```Processor```s
+* Run ```Download``` processor to dump the data sources
+* Run the subroutine: ```FileCentricProcessor``` -> ```IndexProcessor```
+* Run the subroutine: ```ParticipantCentricProcessor``` -> ```IndexProcessor```
+
+The main issue of the above implementation is **lack of flexibility**. 
+
+Consider the following facts:
```

Comment:
Sentences in the bullet list are pretty long (especially the second bullet), It might flow more nicely if those bullets were actually 3 separate paragraphs with a short introductory sentence.